### PR TITLE
[JBPM-9647]  Multiple humantask email notifications are send

### PR DESCRIPTION
**Pull request by fjtirado**
**JIRA**:

[link](https://issues.redhat.com/browse/JBPM-9647)

### CODE REVIEW

File: jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/impl/util/DeadlineSchedulerHelper.java [hunk 1]
```diff
@@ -49,22 +52,32 @@ public static void rescheduleDeadlinesForTask(final InternalTask task, TaskConte
...
         }
-
+        
         if(deadlineTypes.contains(DeadlineType.START)) {
-            for(Deadline deadline : deadlines.getStartDeadlines()) {
-                task.getDeadlines().getStartDeadlines().add(deadline);
+            for(Deadline deadline : task.getDeadlines().getStartDeadlines()) {
+                persistenceContext.remove(deadline);
+            }
+            logger.info("Assigning START deadlines {}", deadlines.getStartDeadlines());
```
Comment (elguardian) [inline jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/impl/util/DeadlineSchedulerHelper.java:60]: too verbose.. switch to debug

File: jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/impl/util/DeadlineSchedulerHelper.java [hunk 2]
```diff
@@ -49,22 +52,32 @@ public static void rescheduleDeadlinesForTask(final InternalTask task, TaskConte
...
                 persistenceContext.persistDeadline(deadline);
             }
         }
         if(deadlineTypes.contains(DeadlineType.END)) {
-            for(Deadline deadline : deadlines.getEndDeadlines()) {
-                task.getDeadlines().getEndDeadlines().add(deadline);
+            for(Deadline deadline : task.getDeadlines().getEndDeadlines()) {
+                persistenceContext.remove(deadline);
+            }  
+            logger.info("Assigning END deadlines {}", deadlines.getEndDeadlines());
```
Comment (elguardian) [inline jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/impl/util/DeadlineSchedulerHelper.java:70]: too verbose.. switch to debug

**Review by elguardian — CHANGES_REQUESTED** [review]
Lack of unit test. otherwise looks good.

**Comment by fjtirado** [comment]
Test with 2 notifications added

File: jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/EmailDeadlinesBaseTest.java [hunk 3]
```diff
@@ -157,6 +157,65 @@ public void testDelayedEmailNotificationOnDeadline() throws Exception {
...
+
+        MimeMessage msg = ((WiserMessage) getWiser().getMessages().get(0)).getMimeMessage();
+        assertEquals(myBody, msg.getContent());
+        assertEquals(mySubject, msg.getSubject());
+        assertEquals("from@domain.com", ((InternetAddress) msg.getFrom()[0]).getAddress());
+        assertEquals("replyTo@domain.com", ((InternetAddress) msg.getReplyTo()[0]).getAddress());
+
+        final Address[] recipients = msg.getRecipients(RecipientType.TO);
+        assertNotNull(recipients);
+        assertTrue(recipients.length == 1);
```
Comment (afalhambra) [inline jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/EmailDeadlinesBaseTest.java:213]: This is always going to be 1 as the number of the recipients in the email notification is always set to 1 for both deadlines.

**Review by afalhambra — CHANGES_REQUESTED** [review]
The fix looks good to me. 
However the unit test added is not completely valid for this scenario. 

I ran this new unit test without applying the fix in this PR and it got successfully executed - it should've failed. 

Please see my comments there regarding the unit test.

File: jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/EmailDeadlinesBaseTest.java (hunk 3 above)
Comment (fjtirado) [reply jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/EmailDeadlinesBaseTest.java:213]: thats the point

File: jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/EmailDeadlinesBaseTest.java [hunk 4]
```diff
@@ -157,6 +157,65 @@ public void testDelayedEmailNotificationOnDeadline() throws Exception {
...
+        List<String> list = new ArrayList<String>(2);
+        list.add(getWiser().getMessages().get(0).getEnvelopeReceiver());
+        list.add(getWiser().getMessages().get(1).getEnvelopeReceiver());
+
+        assertTrue(list.contains("new-email@domain.com"));
+        assertTrue(list.contains("darth@domain.com"));
+
+        MimeMessage msg = ((WiserMessage) getWiser().getMessages().get(0)).getMimeMessage();
+        assertEquals(myBody, msg.getContent());
+        assertEquals(mySubject, msg.getSubject());
```
[the remark this answers is not in the document]
Comment (fjtirado) [reply jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/EmailDeadlinesBaseTest.java:207]: you are right, actually the order is guaranteed by the fact that is a list, so the first one is always scheduled and trigger first

**Comment by fjtirado** [comment]
@elguardian 
Approach changed since the second repeated task was never executed, please re review
@afalhambra 
I think for testing this problem we need a repeated timer example, which is tricky, Ill prefer to merge this one once green and add test later

File: jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/EmailDeadlinesBaseTest.java [hunk 5]
```diff
@@ -141,22 +144,80 @@ public void testDelayedEmailNotificationOnDeadline() throws Exception {
...
+
+        // emails should not be set yet
+        assertEquals(0, getWiser().getMessages().size());
+        
+        countDownListener.waitTillCompleted();
+        
+        for (WiserMessage msg : getWiser().getMessages()) {
+            logger.info(msg.getEnvelopeReceiver());
+        }
+         assertEquals(2, getWiser().getMessages().size());
```
Comment (afalhambra) [inline jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/EmailDeadlinesBaseTest.java:198]:
```suggestion
        assertEquals(2, getWiser().getMessages().size());
```

File: jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/EmailDeadlinesBaseTest.java [hunk 6]
```diff
@@ -141,22 +144,80 @@ public void testDelayedEmailNotificationOnDeadline() throws Exception {
...
         list = new ArrayList<String>(2);
         list.add(((InternetAddress) recipients[0]).getAddress());
         list.add(((InternetAddress) recipients[1]).getAddress());
         assertTrue(list.contains("new-email@domain.com"));
         assertTrue(list.contains("darth@domain.com"));
     }
     
+    
+    @Test(timeout = 10000)
+    public void testDelayedEmail2NotificationsOnDeadline() throws Exception {
```
Comment (afalhambra) [inline jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/EmailDeadlinesBaseTest.java:165]: doesn't throw any checked exception
```suggestion
    public void testDelayedEmail2NotificationsOnDeadline() {  
```

File: jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/EmailDeadlinesBaseTest.java [hunk 7]
```diff
@@ -141,22 +144,80 @@ public void testDelayedEmailNotificationOnDeadline() throws Exception {
...
         assertTrue(list.contains("darth@domain.com"));
     }
     
+    
+    @Test(timeout = 10000)
+    public void testDelayedEmail2NotificationsOnDeadline() throws Exception {  
+        CountDownTaskEventListener countDownListener = new CountDownTaskEventListener(2, false, true);
+        addCountDownListner(countDownListener);
+        
+        Map<String, Object> vars = new HashMap<String, Object>();
```
Comment (afalhambra) [inline jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/EmailDeadlinesBaseTest.java:169]:
```suggestion
        Map<String, Object> vars = new HashMap<>();
```

File: jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/EmailDeadlinesBaseTest.java [hunk 8]
```diff
@@ -141,22 +144,80 @@ public void testDelayedEmailNotificationOnDeadline() throws Exception {
...
+        vars.put("now", new Date());
+
+        Reader reader = new InputStreamReader(getClass().getResourceAsStream(MvelFilePath.DeadlineWith2Notifications));
+        Task task = TaskFactory.evalTask(reader, vars);
+        
+        taskService.addTask(task, new HashMap<String, Object>());
```
Comment (afalhambra) [inline jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/EmailDeadlinesBaseTest.java:175]:
```suggestion
        taskService.addTask(task, new HashMap<>());
```

File: jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/EmailDeadlinesBaseTest.java [hunk 9]
```diff
@@ -141,22 +144,80 @@ public void testDelayedEmailNotificationOnDeadline() throws Exception {
...
+        assertEquals(0, getWiser().getMessages().size());
+        
+        countDownListener.waitTillCompleted();
+        
+        for (WiserMessage msg : getWiser().getMessages()) {
+            logger.info(msg.getEnvelopeReceiver());
+        }
+         assertEquals(2, getWiser().getMessages().size());
+
+        List<String> list = new ArrayList<String>(2);
```
Comment (afalhambra) [inline jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/EmailDeadlinesBaseTest.java:200]:
```suggestion
        List<String> list = new ArrayList<>(2);
```

File: jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/EmailDeadlinesBaseTest.java [hunk 10]
```diff
@@ -141,22 +144,80 @@ public void testDelayedEmailNotificationOnDeadline() throws Exception {
...
+            try {
+                MimeMessage msg = m.getMimeMessage();
+                return "new-email@domain.com".equals(((InternetAddress) msg.getRecipients(RecipientType.TO)[0])
+                        .getAddress()) && myBody.equals(msg.getContent()) && mySubject.equals(msg.getSubject()) &&
+                       "from@domain.com".equals(((InternetAddress) msg.getFrom()[0]).getAddress()) &&
+                       "replyTo@domain.com".equals(((InternetAddress) msg.getReplyTo()[0]).getAddress());
+            } catch (MessagingException | IOException e) {
+                throw new IllegalStateException(e);
+            }
+        }).findAny().isPresent());
```
Comment (afalhambra) [inline jbpm-human-task/jbpm-human-task-core/src/test/java/org/jbpm/services/task/EmailDeadlinesBaseTest.java:217]: you can use `anyMatch` instead, but up to you
```suggestion
        assertTrue("Cannot find expected message", getWiser().getMessages().stream().anyMatch(m -> {
            try {
                MimeMessage msg = m.getMimeMessage();
                return "new-email@domain.com".equals(((InternetAddress) msg.getRecipients(RecipientType.TO)[0])
                        .getAddress()) && myBody.equals(msg.getContent()) && mySubject.equals(msg.getSubject()) &&
                       "from@domain.com".equals(((InternetAddress) msg.getFrom()[0]).getAddress()) &&
                       "replyTo@domain.com".equals(((InternetAddress) msg.getReplyTo()[0]).getAddress());
            } catch (MessagingException | IOException e) {
                throw new IllegalStateException(e);
            }
        }));
```

**Review by afalhambra — APPROVED** [review]
Only a few minor comments, not required. Looks good.
Nice work!

**Comment by afalhambra** [comment]
Looking great to me :)

**Comment by elguardian** [comment]
Jenkins run fdb

**Pull request merged**